### ISSUE TITLE
Jobs suddenly not processing for one out of many queues

### ISSUE THREAD

**Issue by jordi-chacon** [sorentwo/oban#712]
### Environment

- Oban: `2.12.0`
- Oban Pro: `0.11.0`
- PostgreSQL: `PostgreSQL 13.6 on x86_64-pc-linux-gnu, compiled by Debian clang version 12.0.1, 64-bit`
- Elixir: `1.12.2`
- Erlang/OTP: `24.0.3`

### Oban configuration

```
crontab = [
  {"0 1 * * *", ...},
  ...
]

cron_plugin = {Oban.Plugins.Cron, timezone: "America/Los_Angeles", crontab: crontab}

plugins = [
  Oban.Plugins.Stager,
  Oban.Plugins.Gossip,
  {Oban.Plugins.Pruner, [limit: 200_000, max_age: 1200]},
  Oban.Plugins.Reindexer,
  Oban.Pro.Plugins.DynamicLifeline,
  Oban.Pro.Plugins.BatchManager,
  cron_plugin
]

oban_queues = [
  default: [global_limit: 6],
  scheduled: [global_limit: 2],
  fast: [global_limit: 6],
  high_priority: [global_limit: 6],
  q1: [global_limit: 30],
  q2: [global_limit: 30],
  q3: [global_limit: 30],
  ...
]

config :app, Oban,
  engine: Oban.Pro.Queue.SmartEngine,
  repo: App.Repo,
  log: :debug,
  plugins: oban_plugins,
  queues: oban_queues
```

### Current Behavior

Yesterday we noticed that our production environment had not been processing jobs in our busiest queue (let's call it queue `q1`) for the previous 18 hours while all other queues were processing their jobs just fine. We've been running Oban in production for more than 3 years (we couldn't be happier with Oban!) and have never seen this issue before.

#### Debugging

First, we took a look at the `oban_jobs` table and found out that 6.4K jobs inserted 18h ago and scheduled for more than 1h ago had not been yet attempted. They all belonged to the same queue:
```
select queue, count(*)
from oban_jobs
where
  attempted_at is null and
  scheduled_at < now() - interval '1 hour' and
  inserted_at > '2022-05-20 22:00:00'
group by queue;

  queue   | count
----------+-------
   q1     |  6454
```

We then logged in into the `iex` shell of one of our worker nodes and ran `Oban.check_queue(queue: :q1)`:
```
%{
  global_limit: 30,
  local_limit: 30,
  name: "Oban",
  node: "...-worker@...",
  paused: false,
  queue: "q1",
  rate_limit: nil,
  retry_attempts: 5,
  retry_backoff: 1000,
  running: [421294386, 421294393, 421294404, 421294406, 421294414, 421294415,
   421294417, 421294418, 421294419, 421294423, 421294428, 421294432, 421294547,
   421294578],
  started_at: ~U[2022-05-20 09:07:04.610378Z],
  updated_at: ~U[2022-05-21 16:36:56.402003Z]
}
```

Nothing seemed off in the result above at first. However, we noticed the list of `running` jobs would not change as time went by. Running the same command 30 minutes later yielded the exact same `running` list.

We wanted to find out more about those `running` jobs that the `q1` queue was stuck with, so we queried the oldest 1000 jobs that hadn't been attempted yet (I forgot to `SELECT` the `scheduled_at` column, but we _know_ all those jobs are scheduled to be run immediately after being inserted, so you can assume `scheduled_at` to be equal to `inserted_at`):
[oldest_1000_scheduled_but_not_attempted_jobs.txt](https://github.com/sorentwo/oban/files/8749007/oldest_1000_scheduled_but_not_attempted_jobs.txt)

Some observations:
- The first batch of not-attempted jobs, inserted at `2022-05-20 22:00:01.451425`, has priority 2 (low). Their IDs are lower than those `running` in the `q1` queue, suggesting they were inserted earlier.
- The second batch of jobs, inserted at `2022-05-20 22:10:00.789974`, has priority 1. The smallest ID within this batch is `421294176` and the greatest ID is `421294763`. Notice how all IDs in the list of `q1` `running` jobs above (`[421294386, 421294393, 421294404, 421294406, 421294414, 421294415, 421294417, 421294418, 421294419, 421294423, 421294428, 421294432, 421294547, 421294578]`) are greater than the 2nd batch’s smallest ID and less than the 2nd batch’s greatest ID, suggesting these list of `running` jobs belonged to this second batch of jobs.

At this point we had the impression that the `q1` queue had gotten in some corrupted state. It seemed stuck running jobs that didn't exist in the database and wasn't processing any of the more than 6K jobs scheduled within that queue, nor the new jobs being added to that queue every few minutes. We didn't know how to heal it, so we restarted our worker nodes and immediately the jobs in the `q1` queue got processed successfully. Oban has been behaving correctly since then.

We have no clue as to what might have caused this. We believe the root cause might be in Oban itself, but we are definitely not certain.

Thanks!

**Comment by sorentwo** [comment]
The producer keeps updating, which means it's healthy and running. The issue seems to be those running jobs don't "ack" when they've completed or failed.

Can you share a little about the jobs that correspond to those ids? I wonder if they're somehow exiting with a reason or clause that isn't caught by the queue's producer correctly.

**Comment by jordi-chacon** [comment]
The issue happened again this weekend.

> Can you share a little about the jobs that correspond to those ids?

We weren't logging job IDs so I couldn't find any information about those stuck jobs. We just made a patch that logs job IDs for every `[:oban, :job, :start | :stop | :exception]` job. We've also added logging for every `[:oban, :producer, :start | :stop | :exception]` and some `[:oban, :engine, _, _]` events.

> I noticed that you have logs: :debug, do you happen to have log entries for any of those ids?

Actually, we just found out that Oban is running with `log: :debug` in our production environment. We verified by this by running `Oban.config` in our worker nodes. However, no queries are actually being logged. Do you know what might be preventing Oban queries from being logged?

**Comment by jordi-chacon** [comment]
@sorentwo last Friday we discovered the root cause of the problem: we recently deployed a new Oban worker with a faulty implementation which under certain conditions would end up in an infinite loop. As a result, the job would never end. Eventually, these infinitely looping jobs would use up the whole `limit` of the queue, preventing other jobs from being executed. Pretty embarrassing 😅 

I'm really sorry to have bothered you with what turned out to be our own issue, and thank you so much for your assistance!
